src/test/java/com/xwiki/authentication/saml/XWikiGroupManagerTest.java

```diff
@@ -0,0 +1,164 @@
+                (docRef, parameters) -> new DocumentReference("XWiki", "XWikiGroups", docRef);
+
+        private String userName;
+        private String groupName;
+
+        final XWikiMock xwiki;
+        Logger root = (Logger) LoggerFactory.getLogger(org.slf4j.Logger.ROOT_LOGGER_NAME);
+
+        public DSL() {
+            root.setLevel(Level.OFF);
+            context.setWikiId("XWIKI");
+            context.setMainXWiki("MAIN-XWIKI");
+            context.setAction("");
+            context.setResponse(mock(XWikiResponse.class));
+            XWikiRequest request = mock(XWikiRequest.class);
+            HttpSession httpSession = mock(HttpSession.class);
+            when(request.getSession(true)).thenReturn(httpSession);
+            when(request.getSession()).thenReturn(httpSession);
+
+            context.setRequest(request);
+            xwiki = new XWikiMock(context);
+            context.setWiki(xwiki);
+        }
+
+        public DSL user(String userName){
+            this.userName = userName;
+            return this;
+        }
+        public DSL fromGroup(String groupName){
+            this.groupName = groupName;
+            return this;
+        }
+        public ThenDSL whenAddedToGroup(){
+            groupManager = new XWikiGroupManager(currentMixedDocumentReferenceResolver);
+            groupManager.addUserToXWikiGroup(userName,groupName,context);
+            return new ThenDSL();
+        }
+        public ThenDSL whenRemoveFromGroup(){
+            groupManager = new XWikiGroupManager(currentMixedDocumentReferenceResolver);
+            groupManager.removeUserFromXWikiGroup(userName,groupName,context);
+            return new ThenDSL();
+        }
+
+        public DSL userIsAlreadyInGroup(String userName, String groupName) {
+            groupManager = new XWikiGroupManager(currentMixedDocumentReferenceResolver);
+            groupManager.addUserToXWikiGroup(userName,groupName,context);
+            return this;
+        }
+
+        public class ThenDSL {
+            String userName;
+
+            public ThenDSL then() {
+                return this;
+            }
+
+            public ThenDSL user(String userName) {
+                this.userName = userName;
+                return this;
+            }
+
+            public void isInGroup(String groupName) {
+                String fullGroupName = "XWikiGroups."+groupName;
+                XWikiDocument groupDoc =
+                        xwiki.getSavedDocuments().stream().filter(doc -> doc.toString().equals(fullGroupName)).findFirst()
+                                .orElseGet(() -> Assertions.fail("Group " + groupName + " doesn't exist"));
+
+                BaseClass ref = xwiki.getGroupClass(context);
+                BaseObject obj = groupDoc.getXObject(ref.getDocumentReference(), "member", this.userName);
+                assertNotNull(obj, "User is not in the group");
```

Comment:
Make this assert a bit more verbose, when the test errors out, I want to see which user, which group, etc. I'll let you decide which information would be useful.